src/pycapsule.rs

```diff
@@ -0,0 +1,417 @@
+    ///             .unwrap();
+    ///     let context = Box::new(tx);  // `Sender<String>` is our context, box it up and ship it!
+    ///     capsule.set_context(py, Box::into_raw(context) as *mut c_void).unwrap();
+    ///     // This scope will end, causing our destructor to be called...
+    /// });
+    ///
+    /// assert_eq!(rx.recv(), Ok("Destructor called!".to_string()));
+    /// ```
+    pub fn set_context(&self, py: Python, context: *mut c_void) -> PyResult<()> {
+        let result = unsafe { ffi::PyCapsule_SetContext(self.as_ptr(), context) as u8 };
+        if result != 0 {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Gets the current context stored in the capsule. If there is no context, the pointer
+    /// will be null.
+    pub fn get_context(&self, py: Python) -> PyResult<*mut c_void> {
+        let ctx = unsafe { ffi::PyCapsule_GetContext(self.as_ptr()) };
+        if ctx.is_null() && self.is_valid() && PyErr::occurred(py) {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(ctx)
+        }
+    }
+
+    /// Obtains a reference to the value of this capsule.
+    ///
+    /// # Safety
+    ///
+    /// It must be known that this capsule's pointer is to an item of type `T`.
+    pub unsafe fn reference<T>(&self) -> &T {
+        &*(self.get_pointer() as *const T)
+    }
+
+    /// Gets the raw `c_void` pointer to the value in this capsule.
+    pub fn get_pointer(&self) -> *mut c_void {
+        unsafe { ffi::PyCapsule_GetPointer(self.0.as_ptr(), self.name().as_ptr()) }
+    }
+
+    /// Checks if this is a valid capsule.
+    pub fn is_valid(&self) -> bool {
+        let r = unsafe { ffi::PyCapsule_IsValid(self.as_ptr(), self.name().as_ptr()) } as u8;
+        r != 0
+    }
+
+    /// Retrieves the name of this capsule.
+    pub fn name(&self) -> &CStr {
+        unsafe {
+            let ptr = ffi::PyCapsule_GetName(self.as_ptr());
+            CStr::from_ptr(ptr)
+        }
+    }
+}
+
+// C layout, as PyCapsule::get_reference depends on `T` being first.
+#[repr(C)]
+struct CapsuleContents<T: 'static + Send, D: FnOnce(T, *mut c_void)> {
+    value: T,
+    destructor: D,
+}
+
+// Wrapping ffi::PyCapsule_Destructor for a user supplied FnOnce(T) for capsule destructor
+unsafe extern "C" fn capsule_destructor<T: 'static + Send, F: FnOnce(T, *mut c_void)>(
+    capsule: *mut ffi::PyObject,
+) {
+    let ptr = ffi::PyCapsule_GetPointer(capsule, ffi::PyCapsule_GetName(capsule));
+    let ctx = ffi::PyCapsule_GetContext(capsule);
+    let CapsuleContents { value, destructor } = *Box::from_raw(ptr as *mut CapsuleContents<T, F>);
+    destructor(value, ctx)
+}
+
+/// Guarantee `T` is not zero sized at compile time.
+/// credit: https://users.rust-lang.org/t/is-it-possible-to-assert-at-compile-time-that-foo-t-is-not-called-with-a-zst/67685
+pub trait PanicWhenZeroSized: Sized {
```

Comment:
I think this name is a bit nicer:
```suggestion
#[doc(hidden)]
pub trait AssertNotZeroSized: Sized {
```